wrappers/services_test.go

```diff
@@ -257,7 +252,97 @@ WantedBy=multi-user.target
 `,
 		systemd.EscapeUnitNamePath(dir),
 		dirs.GlobalRootDir,
-	))
+	)
+
+	sliceTempl := `[Unit]
+Description=Slice for snap quota group %s
+Before=slices.target
+
+[Slice]
+# Always enable memory accounting otherwise the MemoryMax setting does nothing.
+MemoryAccounting=true
+MemoryMax=%s
+`
+
+	sliceContent := fmt.Sprintf(sliceTempl, grp.Name, memLimit.String())
+
+	exp := []changesObservation{
+		{
+			snapName: "hello-snap",
+			unitType: "service",
+			name:     "svc1",
+			old:      "",
+			new:      svcContent,
+		},
+		{
+			grp:      grp,
+			unitType: "slice",
+			new:      sliceContent,
+			old:      "",
+			name:     "foogroup",
+		},
+	}
+	r, observe := expChangeObserver(c, exp)
+	defer r()
+
+	err = wrappers.EnsureSnapServices(m, nil, observe, progress.Null)
+	c.Assert(err, IsNil)
+	c.Check(s.sysdLog, DeepEquals, [][]string{
+		{"daemon-reload"},
+	})
+
+	c.Assert(svcFile, testutil.FileEquals, svcContent)
+}
+
+type changesObservation struct {
+	snapName string
+	grp      *quota.Group
+	unitType string
+	name     string
+	old      string
+	new      string
+}
+
+func expChangeObserver(c *C, exp []changesObservation) (restore func(), obs wrappers.ObserveChangeCallback) {
+	changesObserved := []changesObservation{}
+	f := func(app *snap.AppInfo, grp *quota.Group, unitType, name, old, new string) {
+		snapName := ""
+		if app != nil {
+			snapName = app.Snap.SnapName()
+		}
+		changesObserved = append(changesObserved, changesObservation{
+			snapName: snapName,
+			grp:      grp,
+			unitType: unitType,
+			name:     name,
+			old:      old,
+			new:      new,
+		})
+	}
+
+	r := func() {
+		// sort the changesObserved by snapName, with all services being
+		// observed first, then all groups secondly
+		groupObservations := make([]changesObservation, 0, len(changesObserved))
+		svcObservations := make([]changesObservation, 0, len(changesObserved))
+
+		for _, chg := range changesObserved {
+			if chg.unitType == "slice" {
+				groupObservations = append(groupObservations, chg)
+			} else {
+				svcObservations = append(svcObservations, chg)
+			}
+		}
+		sort.SliceStable(svcObservations, func(i, j int) bool {
```

Comment:
don't we need to sort groupObservations too?

Reply:
no, since we already do that in AllQuotaGroups - when we call EnsureSnapServices, the order in which we iterate over the groups is determined by AllQuotaGroups which already does sort.SliceStable on the groups first.

Reply:
maybe a comment would be good for future readers:

// groups are sorted already as the code uses QuotaGroupSet.AllQuotaGroups

Reply:
added a comment